refactor: tooltip to function component

ckifer opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

<!--- Describe your changes in detail -->
Refactor `Tooltip` to be a function component

## Related Issue

<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->
Refactoring library to be more maintainable - testing refactorability

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):
https://user-images.githubusercontent.com/25180830/233861482-13dd6acb-c8cc-4058-99e5-cf3c8231fd79.mov

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

ckifer commented:
@nikolasrieble - did this as a draft for fun, definitely not done or tested. Not sure if we want to take tasks like this yet but oh well.

I actually think I notice some performance improvement over the class based approach but I'm not 100% sure. Feel free to pull down and play

nikolasrieble commented:
Amazing, will test it today.

ckifer commented on test/chart/LineChart.spec.tsx:
```diff
@@ -648,7 +648,7 @@ describe('<LineChart /> - Rendering two line charts with syncId', () => {
       </LineChart>
     );
-    const { container, getByText } = render(
+    const { container, getByText, debug } = render(
```
oops, left these here. Not planning on merging this so will leave them for now

ckifer commented on storybook/stories/API/chart/BarChart.stories.tsx:
```diff
@@ -12,6 +12,7 @@ export const Simple = {
       <ResponsiveContainer width="100%" height={400}>
         <BarChart {...args}>
           <Bar dataKey="uv" />
+          <Tooltip reverseDirection={{ x: true }} />
```
same with this, just testing. Would remove in non-draft

ckifer commented:
I mean it works 😅  - probably not a priority rn though

nikolasrieble commented:
> I mean it works 😅 - probably not a priority rn though

Well, a refactoring is always important. I will test this thoroughly and then lets see whether the behaviour replicates.

nikolasrieble commented on src/component/Tooltip.tsx:
```diff
@@ -202,101 +196,120 @@ export class Tooltip<TValue extends ValueType, TName extends NameType> extends P
+ */
+Tooltip.defaultProps = tooltipDefaultProps;
+
+export { Tooltip };
```
Nit: Why export it in an object? Why not simply `export Tooltip`?

ckifer replied:
oh, habit I guess -  I'm used to exporting multiple things at the bottom of a file

nikolasrieble commented on src/component/Tooltip.tsx:
```diff
@@ -202,101 +196,120 @@ export class Tooltip<TValue extends ValueType, TName extends NameType> extends P
+      }}
+      className={cls}
+      style={outerStyle}
+      ref={wrapperNode}
```
I am not sure whether this change equivalent.

ckifer replied:
Not sure either - this is the approach I would take with refs in general (and how they should be done now), but might be a breaking change. Will test to see if it is. Can revert to a legacy ref if needed

nikolasrieble commented on src/component/Tooltip.tsx:
```diff
@@ -202,101 +196,120 @@ export class Tooltip<TValue extends ValueType, TName extends NameType> extends P
-      // https://github.com/recharts/recharts/pull/2925
-      // eslint-disable-next-line jsx-a11y/no-noninteractive-element-interactions
-      <div
-        tabIndex={-1}
```
Why remove the tabIndex?

ckifer replied:
This should be in a separate PR sorry - but imho it shouldn't be there. It was added with accessibility in mind and its only made accessibility worse by blurring focused elements (say in a form) on chart hover.

Will add it back for this PRs sake

nikolasrieble commented on src/component/Tooltip.tsx:
```diff
@@ -90,82 +90,78 @@ export type TooltipProps<TValue extends ValueType, TName extends NameType> = Def
+const Tooltip = <TValue extends ValueType, TName extends NameType>(
+  props: TooltipProps<TValue, TName> & { children?: React.ReactNode },
+) => {
+  const [state, setState] = useState({
```
While this works, it is recommended to split the state into separate variables. 
https://reactjs.org/docs/hooks-faq.html#should-i-use-one-or-many-state-variables

ckifer replied:
agreed, was doing this for the sake of speed when originally refactoring. Can revise

nikolasrieble commented on src/component/Tooltip.tsx:
```diff
@@ -90,82 +90,78 @@ export type TooltipProps<TValue extends ValueType, TName extends NameType> = Def
+  allowEscapeViewBox: { x: false, y: false },
+  reverseDirection: { x: false, y: false },
+  offset: 10,
+  viewBox: { x: 100, y: 100, height: 50, width: 50 },
```
What is the reasoning for these specific values? - as opposed to the existing ones?

ckifer replied:
`{ x1: 0, x2: 0, y1: 0, y2: 0 }` - these are not valid values on `viewBox` - there is no such thing as x1, y1, x2, y2

As for the values I think these are the default viewBox values for the tooltip but I honestly forget. Will check

nikolasrieble replied:
I understand the change, but it seems an unrelated bug fix. I did not find these default values anywhere, I believe this might be changing behaviour. Please consider doing this change in a prefactoring PR, keeping the refactoring class --> function a pure refactoring without behavioural change.

nikolasrieble replied:
I will test the Tooltip behaviour on this PR. This is the only change that I suspect might be an issue / change behaviour.

ckifer replied:
Hmm okay, I think I got these values from somewhere but I don't remember exactly where. Either way these x1, y1 etc. fields are incorrect and need addressed somewhere. I can ts-ignore them here and fix later

nikolasrieble commented on src/component/Tooltip.tsx:
```diff
@@ -90,82 +90,78 @@ export type TooltipProps<TValue extends ValueType, TName extends NameType> = Def
-    contentStyle: {},
-    itemStyle: {},
-    labelStyle: {},
-    cursor: true,
-    trigger: 'hover',
```
Why remove these two?

ckifer replied:
which two?
`cursor: true,` is still there

`trigger: 'hover',` - I forget, might just be a miss (though it does get set without this defaulted)

will add it back

nikolasrieble commented on src/component/Tooltip.tsx:
```diff
@@ -90,82 +90,78 @@ export type TooltipProps<TValue extends ValueType, TName extends NameType> = Def
-  }
+  };
+
+  useEffect(() => {
+    updateBBox();
+  }, []);
```
Previously `updateBBox()` was called both on onComponentMount as well as onComponentUpdate.

This is not equivalent. The useEffect will only fire onMount, but not onUpdate. 
If the behaviour should replicate, then state would have to be added into the dependency array.

nikolasrieble reviewed (commented):
Sorry that the review took so long, I was heads down in storybook. Have a few questions.

ckifer reviewed (commented):
Thanks for the review, will make some changes

ckifer commented:
@nikolasrieble I think I addressed all comments here, this _should_ be ready for re-review and testing. On the storybook all looks good to me. 

I'm going to leave the `useRef` change, I'm confident that the changes are equivalent

nikolasrieble commented on src/chart/generateCategoricalChart.tsx:
```diff
@@ -2157,6 +2157,7 @@ export const generateCategoricalChart = ({
       }
 
       const { children, className, width, height, style, compact, title, desc, ...others } = this.props;
+
```
```suggestion
```
Unrelated change.

ckifer replied:
oh sorry, unintentional

nikolasrieble commented on src/component/Tooltip.tsx:
```diff
@@ -90,99 +99,97 @@ export type TooltipProps<TValue extends ValueType, TName extends NameType> = Def
+  offset: 10,
+  viewBox: { x: 100, y: 100, height: 50, width: 50 },
+  coordinate: { x: 0, y: 0 },
+  wrapperStyle: {},
+  // this doesn't exist on TooltipProps
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore
+  cursorStyle: {},
+  separator: ' : ',
+  contentStyle: {},
+  itemStyle: {},
+  labelStyle: {},
+  isAnimationActive: !Global.isSsr,
+  animationEasing: 'ease',
+  animationDuration: 400,
+  filterNull: true,
+  useTranslate3d: false,
+  cursor: true,
+  trigger: 'hover',
```
Nit: Any reason these a ordered differently now? If you want to change the order, alphabetically might be nice. Else keeping a PR as small as possible (i.e. no redundant changes) would set a great example.

ckifer replied:
No reason, just how it happened. Will move back to original order

nikolasrieble commented on src/util/ReactUtils.ts:
```diff
@@ -416,8 +416,8 @@ export const renderByOrder = (children: React.ReactElement[], renderMap: any) =>
     } else if (child) {
       const displayName = getDisplayName(child.type);
-      const { handler, once } = renderMap[displayName] || {};
 
+      const { handler, once } = renderMap[displayName] || {};
```
Nit: Ideally no redundant changes.

ckifer reviewed (commented):
Will put out one more revision

ckifer commented on src/component/Tooltip.tsx:
```diff
@@ -90,99 +99,100 @@ export type TooltipProps<TValue extends ValueType, TName extends NameType> = Def
-      });
-    }
-  };
+const tooltipDefaultProps: TooltipProps<number, string> = {
```
typed as number, string for default props. Doesn't really affect anything except for typing this explicitly

ckifer commented on src/component/Tooltip.tsx:
```diff
@@ -90,99 +99,100 @@ export type TooltipProps<TValue extends ValueType, TName extends NameType> = Def
+  // TODO: replace invalid properties
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore
+  viewBox: { x1: 0, x2: 0, y1: 0, y2: 0 },
```
these are incorrect default values - fix for this out of scope of PR

ckifer reviewed (commented):
@nikolasrieble requested changes made. defaultProps back in original order and original content. Check out the video in the PR description for a glance at functionality without having to pull the branch down

nikolasrieble reviewed (approved):
Great, thank you for addressing all comments! 🚀 Would you like to create a follow up issue for the invalid default values of the viewBox?

ckifer commented:
https://github.com/recharts/recharts/issues/3548

ckifer commented:
Lets test this in the storybook deployment
